pkg/namespace-labeler/labeler.go

```diff
@@ -29,25 +31,41 @@ import (
 type Labeler struct {
 	enforcedNsLabel       string
 	prometheusRuleLabeler bool
-	excludeList           map[string]map[string]struct{}
+	excludeList           map[string]map[string][]string
+	//deprecated - keep for backward compatibility
+	excludedRulesConfig map[string]map[string]struct{}
 }
 
 // New - creates new Labeler
 // enforcedNsLabel - label name to be enforced for namespace
-// excludeConfig - list of namespace + PrometheusRule names to be excluded while enforcing adding namespace label
-func New(enforcedNsLabel string, excludeConfig []monitoringv1.PrometheusRuleExcludeConfig, prometheusRuleLabeler bool) *Labeler {
+// excludeConfig - list of ObjectReference to be excluded while enforcing adding namespace label
+// prometheusRuleLabeler - whether this should apply for Prometheus or Thanos rules
+// excludedRulesConfig - deprecated, rule exclusion rules. Use excludeConfig
+func New(enforcedNsLabel string, excludeConfig []monitoringv1.ObjectReference, prometheusRuleLabeler bool,
+	excludedRulesConfig []monitoringv1.PrometheusRuleExcludeConfig) *Labeler {
 
 	if enforcedNsLabel == "" {
 		return &Labeler{} // no-op labeler
 	}
 
-	if len(excludeConfig) == 0 {
+	if len(excludeConfig) == 0 && len(excludedRulesConfig) == 0 {
 		return &Labeler{enforcedNsLabel: enforcedNsLabel}
 	}
 
-	ruleExcludeList := make(map[string]map[string]struct{})
-
+	objectsExcludeList := make(map[string]map[string][]string)
 	for _, r := range excludeConfig {
+		if r.Namespace == "" || r.GroupResource().Empty() {
```

Comment:
this shouldn't be possible given the validations at the OpenAPI spec level.

Reply:
defensive checks for tests